File: python/mirage/commands/builtin/generic_bind/builders/ls.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import Awaitable, Callable
from functools import partial

from mirage.accessor.base import Accessor
from mirage.commands.builtin.generic.ls import ls_generic
from mirage.commands.builtin.generic_bind.adapter import (Builder, CommandIO,
                                                          overlaid_stat)
from mirage.commands.config import CommandOpts
from mirage.io.types import ByteSource, IOResult
from mirage.types import FileStat, PathSpec


async def ls(ops: CommandIO, accessor: Accessor, paths: list[PathSpec],
             texts: list[str],
             opts: CommandOpts) -> tuple[ByteSource | None, IOResult]:
    if not ops.is_mounted(accessor):
        raise ValueError("ls: no resource")
    if not paths:
        cwd_virtual = opts.cwd.virtual if isinstance(opts.cwd,
                                                     PathSpec) else opts.cwd
        cwd_rp = (opts.cwd.resource_path
                  if isinstance(opts.cwd, PathSpec) else opts.cwd.strip("/"))
        paths = [
            PathSpec(virtual=cwd_virtual,
                     directory=cwd_virtual,
                     resolved=False,
                     resource_path=cwd_rp)
        ]
    resolved = await ops.resolve_glob(accessor, paths, opts.index)
    stat_fn: Callable[..., Awaitable[FileStat]] = partial(ops.stat, accessor)
    if opts.stat_overlay is not None:
        stat_fn = partial(overlaid_stat, stat_fn, opts.stat_overlay)
    return await ls_generic(resolved, list(texts), opts,
                            partial(ops.readdir, accessor), stat_fn)


BUILDER = Builder('ls', ls, None, False, None)

File: typescript/packages/core/src/commands/builtin/generic_bind/builders/ls.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { PathSpec } from '../../../../types.ts'
import { lsGeneric } from '../../generic/ls.ts'
import { type Builder, resolveGlobOf } from '../adapter.ts'

export const LS_BUILDER: Builder = {
  name: 'ls',
  fn: async (ops, accessor, paths, _texts, opts) => {
    const idx = opts.index ?? undefined
    const resolved = paths.length > 0 ? await resolveGlobOf(ops)(accessor, paths, idx) : []
    const overlay = opts.statOverlay
    // ls renders stat rows the backend produces, which never see the
    // namespace attr overlay (chmod/chown/touch on overlay backends);
    // merge it in so ls -l matches the ops facade.
    const stat =
      overlay !== undefined
        ? async (p: PathSpec) => overlay(p.virtual, await ops.stat(accessor, p, idx))
        : (p: PathSpec) => ops.stat(accessor, p, idx)
    return lsGeneric(resolved, opts, (p) => ops.readdir(accessor, p, idx), stat)
  },
}
